### PR TITLE
Issue#336 Stop the cohorts_validation using break from within a validation proc

### PR DESCRIPTION
**Pull request by blackrat**
### Context

Fixes issue #336
Breaks the coverage tests since a break inside a proc is an illegal instruction.

### Changes proposed in this pull request

- Switching each to any? and dropping the break will stop after the first error is added, which was the intent.

errors.add returns the errors object, which is truthy, so will stop the loop and only add it once.

### Review Checks
- [ ] All pages have automated accessibility checks via cypress
- [ ] All pages have visual tests via cypress + percy
- [ ] All `School` queries are correctly scoped - `eligible` when they need to be

### CODE REVIEW

File: app/forms/delivery_partner_form.rb [hunk 1]
```diff
@@ -86,10 +86,9 @@ def lead_providers_validation
   def cohorts_validation
     # Ensure all selected lead providers have at least one selected cohort
     # This is indicated by the presence of a provider relationship for that lead provider
-    lead_provider_ids&.filter(&:present?)&.each do |lead_provider_id|
+    lead_provider_ids&.filter(&:present?)&.any? do |lead_provider_id|
       unless chosen_provider_relationships.pluck(:lead_provider_id).include?(lead_provider_id)
         errors.add(:provider_relationship_hashes, :blank, message: "Choose at least one cohort for every lead provider")
-        break
```
Comment (kicferk1) [inline app/forms/delivery_partner_form.rb:92]: What happens if we have have two `lead_provider_ids `?

Do we get duplicate error?

Comment (kicferk1) [reply app/forms/delivery_partner_form.rb:92]: Also, in principle I dislike side effects in functions like `map`, `filter`, and yes, `any` - `each` is the 'principled' functional array method that is designed to allow for those.

Maybe a slightly more semantically correct way of going about it would be
```
cohorts_selected =  lead_provider_ids&.filter(&:present?)&.all? do | lead_provider_id | 
  chosen_provider_relationships.pluck(:lead_provider_id).include?(lead_provider_id)}
end

errors.add(:provider_relationship_hashes, :blank, message: "Choose at least one cohort for every lead provider") unless cohorts_selected
```

Thoughts?

Comment (ethanmills) [reply app/forms/delivery_partner_form.rb:92]: > Do we get duplicate error?
I would assume that `any?` short-circuits and so prevents that. I prefer your suggestion, however

Comment (blackrat) [reply app/forms/delivery_partner_form.rb:92]: The former stops as soon as it hits the condition, the latter runs them all. I prefer the former as closer to the original. 
If you wanted to have a "true" at the end of the block to make it less of a side-effect.

Comment (timabell) [reply app/forms/delivery_partner_form.rb:92]: I can see the argument for `any` being more of a functional list/comprehension type thing so having procedural stuff in it being a bit odd, but given it currently breaks the coverage I'd probably take it as is and then polish later if we feel like it.

The use of `all` will likely be less performant than `any` as it has to run for the whole set rather than exit early.

Comment (timabell) [reply app/forms/delivery_partner_form.rb:92]: how about using `first` and then checking for whether that returned anything....

Comment (timabell) [reply app/forms/delivery_partner_form.rb:92]: I think the use of `errors.add` to generate a `true` for the `unless` is a bit hard to follow, I'll try and make something I like better....

Comment (kicferk1) [reply app/forms/delivery_partner_form.rb:92]: Performance would be my last concern here, we are talking about order of magnitude of '6 lead providers' 😂

Comment (timabell) [reply app/forms/delivery_partner_form.rb:92]: Excuse the deleted comments, I had to play with the code to really grok this.

I agree with @kicferk1 on the suggested change, sorry @blackrat !

Comment (blackrat) [reply app/forms/delivery_partner_form.rb:92]: I dislike the variable usage. I'm more inclined to pull that into a different method and refer to it that way. Thought?

Comment (kicferk1) [reply app/forms/delivery_partner_form.rb:92]: I don't mind putting the code checking if there is a cohort selected for each LP into its own method.

Comment (blackrat) [reply app/forms/delivery_partner_form.rb]: Actually, I had some time to put more thought into this. It was a quick first thing that worked with minimal code, but in thinking about it, it just comes down to an array comparison. If there are any elements in the first array that aren't in the second element, then we have an error. i.e. if (A-B) is not empty add the error in, otherwise it's fine.

File: app/forms/delivery_partner_form.rb [hunk 2]
```diff
@@ -86,10 +86,10 @@ def lead_providers_validation
...
     # Ensure all selected lead providers have at least one selected cohort
     # This is indicated by the presence of a provider relationship for that lead provider
-    lead_provider_ids&.filter(&:present?)&.any? do |lead_provider_id|
-      unless chosen_provider_relationships.pluck(:lead_provider_id).include?(lead_provider_id)
-        errors.add(:provider_relationship_hashes, :blank, message: "Choose at least one cohort for every lead provider")
-      end
+    # Boils down to a set comparison of Arrays of lead_provider_ids
+    # i.e. (A-B) returns all elements in A not included in B, so if there are any, that a validation failure.
+    if (lead_provider_ids&.filter(&:present?)-chosen_provider_relationships.pluck(:lead_provider_id)).any?
+      errors.add(:provider_relationship_hashes, :blank, message: "Choose at least one cohort for every lead provider")
```
Comment (kicferk1) [inline app/forms/delivery_partner_form.rb:92]: There are now 4 lines of code comments here, which I doubt anyone will read in full.

I reckon introducing some local variables with expressive names will do more for code maintainability than code comments.
```
selected_lead_provider_ids = lead_provider_ids&.filter(&:present?)
lead_providers_ids_with_cohort = leachosen_provider_relationships.pluck(:lead_provider_id)
if (selected_lead_provider_ids - lead_providers_ids_with_cohort).any?
  ...
```
That said, I think we are spending more time on it than it's worth :P

**Review by kicferk1 — APPROVED** [review]

File: app/forms/delivery_partner_form.rb (hunk 2 above)
Comment (blackrat) [reply app/forms/delivery_partner_form.rb]: I've extracted the `lead_provider_ids&filter(&:present?)`into a method since it's used in more than one place. The new comment was more to show the rationale behind it, and I'm not sure how valuable the original comment is. My initial thought was to extract into a method with a name close to that comment description. You're right that we're spending a long time over a tiny issue though. Working code replacing broken code was the desire here.

**Review by kicferk1 — APPROVED** [review]

**Review by timabell — APPROVED** [review]

File: app/forms/delivery_partner_form.rb (hunk 1 above)
Comment (ethanmills) [reply app/forms/delivery_partner_form.rb]: Set comparison, perhaps, but it's possible to have multiple copies of the lead provider in chosen_provider_relationships
